libsolidity/ast/Types.h

```diff
@@ -398,10 +398,36 @@ class Type
 	mutable std::optional<size_t> m_stackSize;
 };
 
+/**
+ * Base class for Types that can be explicitly converted from another type, which can potentially
+ * lead to a loss in precision.
+ */
+class LossfullyConvertible
+{
+protected:
+	BoolResult lossfullyConvertible(Type const& _from, Type const& _to) const
+	{
+		if (m_lossfullyConverted && (_from != _to))
+			return BoolResult::err(
+				"The type \"" +
+				_from.toString() +
+				"\" already underwent a potentially lossfull conversion."
+			);
+
+		if (auto lossfullyConvertible = dynamic_cast<LossfullyConvertible const*>(&_to))
+			if (_from != _to)
+				lossfullyConvertible->m_lossfullyConverted = true;
+
+		return true;
+	}
+
+	bool mutable m_lossfullyConverted = false;
```

Comment:
Not sure if the choice of mutable is right here. Looking for suggestions.

Reply:
I find it unintuitive that simply calling `lossfullyConvertible()` (which is a `const` function so it looks harmless) can mark the type as converted, even if no conversion really took place.

Also, what if two values refer to the same type instance? I'd expect something like this on a value, not on its type.

Just a quick idea: maybe the flag should be better as an annotation in the AST instead?

Reply:
> Just a quick idea: maybe this flag would work better as an annotation on expressions in the AST?

Do you think this would work in the following case:

```solidity
		uint x = uint(int(-1));
		uint8 y = uint8(x);
```

Currently it would produce the following error:
> Error: Explicit type conversion not allowed from "uint256" to "uint8". The type "uint256" already underwent a potentially lossfull conversion.
> --> /tmp/example.sol:4:13:
> |
> 4| 		uint8 y = uint8(x);
> | 		          ^^^^^^^^

Reply:
Let me check if annotating expressions would work in this case.

Reply:
But why would you want an error here? `uint8(uint(int(-1)))` would work so it should work if you split it too. There are two separate explicit conversions here.